app/Community/Services/SubscriptionService.php

```diff
@@ -57,58 +62,205 @@ public function isSubscribed(User $user, SubscriptionSubjectType $subjectType, i
-     * @return Builder<Model>
+     * @return Collection<int, Subscription>
+     */
+    public function getSubscriptions(User $user, array $subjectTypes, int $offset = 0, int $count = 100): Collection
+    {
+        $results = new Collection();
+
+        foreach ($subjectTypes as $subjectType) {
+            $subscriptionIds = $this->getSubscriptionSubjectIds($user, $subjectType);
+            $subscriptionCount = count($subscriptionIds);
+
+            if ($offset < $subscriptionCount) {
+                $subscriptions = $this->getSubjectSubscriptions($user, $subjectType, $subscriptionIds, $offset, $count);
+                $results = $results->merge($subscriptions);
+
+                $count -= ($subscriptionCount - $offset);
+                if ($count <= 0) {
+                    break;
+                }
+
+                $offset = 0;
+            } else {
+                $offset -= $subscriptionCount;
+            }
+        }
+
+        return $results;
+    }
+
+    /**
+     * @return Collection<int, Subscription>
      */
-    private function getImplicitSubscriptionsQuery(SubscriptionSubjectType $subjectType, ?int $subjectId, ?int $forUserId = null, ?array $ignoreUserIds = null): Builder
+    private function getSubjectSubscriptions(User $user, SubscriptionSubjectType $subjectType, array $subjectIds, int $offset, int $count): Collection
+    {
+        $handler = $this->getHandler($subjectType);
+        $subjects = $handler->getSubjectQuery($subjectIds)
+            ->toBase() // tell Eloquent not to bind to whatever model was used to construct the query
+            ->offset($offset)
+            ->limit($count)
+            ->get();
+        $resultIds = $subjects->pluck('subject_id')->toArray();
+
+        $explicitSubscriptions = Subscription::query()
+            ->where('user_id', $user->id)
+            ->where('subject_type', $subjectType)
+            ->whereIn('subject_id', $resultIds)
+            ->get()
+            ->mapWithKeys(fn ($item) => [$item['subject_id'] => $item]);
+
+        $results = new Collection();
+        foreach ($subjects as $row) {
+            if ($explicitSubscriptions->has($row->subject_id)) {
+                $result = $explicitSubscriptions[$row->subject_id];
+            } else {
+                $result = new Subscription([
+                    'user_id' => $user->id,
+                    'subject_type' => $subjectType,
+                    'subject_id' => $row->subject_id,
+                    'state' => true,
+                ]);
+            }
+
+            $result->title = $row->title;
+
+            $results->push($result);
+        }
+
+        return $results;
+    }
+
+    private function getSubscriptionSubjectIds(User $user, SubscriptionSubjectType $subjectType): array
+    {
+        $ttl = 12 * 60 * 60; // 12 hours
+
+        return Cache::remember("users.{$user->id}.subscriptions.{$subjectType->value}", $ttl, function () use ($user, $subjectType) {
```

Comment:
I'm assuming the UI will need to query the total number of results in addition to multiple slices of results as the user goes from page to page. This is an attempt to only do the implicit subscription calculations once for all the little interactions. It might be premature optimization, and could potentially be removed once the UI is in place.